FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+pub fn self_destruct(addr_off: *const u8, addr_len: u32) -> Result<()>;
+```
+
+#### Namespace: `vm`
+
+```rust
+/// Abort execution with the given code and message. The code is recorded in the receipt, the
+/// message is for debugging only.
+pub fn abort(code: u32, message: *const u8, message_len: u32) -> !;
+```
+
+### Externs
+
+```rust
+/// Consensus related methods.
+pub trait Consensus {
+    /// Verify a consensus fault.
+    fn verify_consensus_fault(
+        &self,
+        h1: &[u8],
+        h2: &[u8],
+        extra: &[u8],
+    ) -> Result<Option<ConsensusFault>>;
+}
+
+/// Randomness provider trait
+pub trait Rand {
+    /// Gets 32 bytes of randomness for ChainRand paramaterized by the DomainSeparationTag,
+    /// ChainEpoch, Entropy from the ticket chain.
+    fn get_chain_randomness(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+
+    fn get_chain_randomness_looking_forward(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+
+    /// Gets 32 bytes of randomness for ChainRand paramaterized by the DomainSeparationTag,
+    /// ChainEpoch, Entropy from the latest beacon entry.
+    fn get_beacon_randomness(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+
+    fn get_beacon_randomness_looking_forward(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+}
+
+```
+
+### IPLD memory model
+
+IPLD is the data model of Filecoin. Actors can be construed as logic that
+receives an input IPLD graph, performs computation, and (optionally) mutates the
+IPLD graph by saving an alternative root in the actor's state.
+
+All state will be stored and retrieved as IPLD. The state store lives on the
+node side, and it is exposed to the FVM via an Extern. This is important as the
+VM itself needs to understand the IPLD so we can do garbage collection, etc.
+
+IPLD codecs will be defined as WASM libraries, imported through the WASM import
+system. Ideally, ADLs and selectors would be implemented in the same way.
+
+The FVM must guarantee that actors can only access data that is in their state
+tree. This is done through the maintenance of an "accessible set" inside the
+kernel.
+
+Because Filecoin IPLD state objects are highly atomized and linked, accessing
+and mutating entries in objects like HAMTs and AMT results in multiple
+sequential state IO operations, each of which traverses the Extern boundary in a
+non-parallelizable way. If the Extern is traversed through FFI, the cost of
+operating on ADLs may be non-negligible.
+
+Alternative strategies like moving ownership of the state store to the FVM-side,
+or leveraging selectors for queries, may be explored in the future.
+
+### Error conditions
```

Comment:
I'd drop this. We can mention that syscalls have "error numbers" in the syscall section, and message exit codes are already defined.